GameOfRunes/GameOfRunes/Levels and Waves/LevelCreator.swift

```diff
@@ -0,0 +1,259 @@
+                                                              monsterDifficulties: availableMonsterDifficulties,
+                                                              bucket: bucket)
+
+        var level = EnemySpawnUnit()
+        for spawnWaveDifficulty in chunkedDifficulties {
+            let spawnWaveMonsters = LevelCreator.convertDifficultiesToMonsters(difficulties: spawnWaveDifficulty)
+            do {
+                let enemyUnit = try EnemySpawnUnit(spawnWaveMonsters)
+                level += enemyUnit
+            } catch {
+                fatalError("Unable to create level. Cannot proceed")
+            }
+        }
+
+        return level
+    }
+
+    /**
+     Helper function to convert an array of monsters to an array of their associated difficulties.
+     */
+    static func convertMonsterToDifficulties(monsters: [EnemyType]) -> [Int] {
+        var monsterDifficulties: [Int] = []
+        for monster in monsters {
+            assert(monsterToDifficultyMap[monster] != nil)
+            if let monsterDifficulty = monsterToDifficultyMap[monster] {
+                monsterDifficulties.append(monsterDifficulty)
+            }
+        }
+        monsterDifficulties.sort()
+        return monsterDifficulties
+    }
+
+    /**
+     Helper function to convert an array of monster difficulties to an array of monsters.
+     */
+    static func convertDifficultiesToMonsters(difficulties: [Int]) -> [EnemyType] {
+        var monsters: [EnemyType] = []
+        for difficulty in difficulties {
+            assert(difficultyToMonsterMap[difficulty] != nil)
+            if let monster = difficultyToMonsterMap[difficulty] {
+                monsters.append(monster)
+            }
+        }
+        return monsters
+    }
+
+    /**
+     Creates the probability demarcation for each monster difficulty.
+     - Note: The bucket is used to determine which monster difficulty should be selected
+     after sampling from a normal distribution. Hence, the demarcation determines the relative
+     proportion of each monster difficulty in the resulting sample.
+     */
+    private static func createProbabilityBuckets(availableMonsterDifficulties: [Int]) -> [Double] {
+        let totalNumMonsters = Double(availableMonsterDifficulties.count)
+        let denominator: Double = 0.5 * (totalNumMonsters) * (totalNumMonsters + 1)
+
+        var numerator: Double = 0
+        var bucket: [Double] = []
+
+        for i in (1...availableMonsterDifficulties.count).reversed() {
+            numerator += Double(i)
+            bucket.append(numerator / denominator)
+        }
+
+        return bucket
+    }
+
+    /**
+     Gets the index of the bucket in which `probability` lies.
+     */
+    private static func getProbabilityBucketIndex(bucket: [Double], probability: Double) -> Int {
+        let convenienceBucket = [0] + bucket
+        for i in (1...(convenienceBucket.count - 1)) {
```

Comment:
NIT: Can use `..<convenienceBucket.count` instead of ...(convenienceBucket.count. -1)

Reply:
refactored